docs/DesignDocs/TEECallsRefactoring.md

```diff
@@ -0,0 +1,519 @@
+Each enclave defines its own EDL file with a different number of ECALLs.
+In addition, both the EDL files import the same EDL file (the same `sample.edl` as above).
+As a result, the corresponding ECALL tables (defined in each `_t.c` file) include
+the wrapper function of `sample_ecall` but with different function id (because
+the number of ECALLs is different).
+Assuming that we use *weak symbol* to avoid the naming conflict of
+the wrapper function (i.e., only one implementation of the functions is picked),
+the invocation of `sample_ecall` to both enclaves ends up using the same
+table id and function id and potentially causing a mismatching;
+i.e., the enclave may use the wrong table id or function id to look up the ECALL.
+
+Proposed revision
+-----------
+
+To address two limitations more fundamentally, this document proposes to
+use a single table that includes ECALLs or OCALLs in multiple EDL files.
+This section describes the detail of the design and considerations.
+
+**Maintain Single Table for ECALL and OCALL**
+
+Instead of adopting the per-EDL tables, maintaining a single table (i.e., one
+OCALL table and one ECALL table) provides finer granularity to keep track
+of ECALLs or OCALLs from multiple EDL files that are either imported or directly included.
+To avoid duplicated entries, each ECALL or OCALL is registered by its hash of the name.
+  
+**Add the registration functions of ECALLs and OCALLs in `_t.c` and `_u.c`**
+
+Instead of using wrappers for `_t.c` and `_u.c` files, each `_t.c` or `_u.c` now
+includes functions that register the list of OCALLs/ECALLs by passing the function
+pointers and the hashes of function names. See the following example.
+  
+```
+static oe_ecall_func_t __oe_ecalls_table[] = {
+    (oe_ecall_func_t) ecall_oe_verify_report_ecall,
+    (oe_ecall_func_t) ecall_oe_get_public_key_ecall,
+    (oe_ecall_func_t) ecall_oe_get_public_key_by_policy_ecall,
+    (oe_ecall_func_t) ecall_oe_log_init_ecall
+};
+
+static uint32_t __oe_ecalls_hash_list[] = {
+    3802944385 /* CRC32 of oe_verify_report_ecall. */,
+    1967956880 /* CRC32 of oe_get_public_key_ecall. */,
+    550431498 /* CRC32 of oe_get_public_key_by_policy_ecall. */,
+    4072068939 /* CRC32 of oe_log_init_ecall. */
+};
+
+static size_t __oe_ecalls_table_size = OE_COUNTOF(__oe_ecalls_table);
+
+/* Ecall table registration function. */
+static void register_core_ecall_functions()
+{
+    oe_register_ecall_functions(
+      __oe_ecalls_table,
+      __oe_ecalls_hash_list,
+      __oe_ecalls_table_size);
+}
+```
+
+Each table entry is then defined as the following struct.
+
+```
+typedef struct _oe_ocall_t {
+    oe_ocall_func_t ocall;
+    uint32_t hash;
+} oe_ocall_t;
+```
+
+**Maintain the List of Registration Functions**
+
+Each registration function in a EDL file is added to a list via the constructor.
+On Linux, we can do this via ` __attribute__((constructor))` as follows.
```

Comment:
One issue with `__attribute__((constructor))` is that we cannot guarantee the order in which these constructors will be called. What if the user code has an constructor that makes an ocall?

Can we ensure that we invoke these constructors at a well defined point in enclave initialization?

Reply:
We don't do the initialization inside the constructor. Instead, it simply adds the registration functions to a linked list.  The list of the functions will be invoked at a well-defined point (the same as how we register the per-EDL table today). It will be clearer when I create the patch :)

Reply:
The text is not correct.   `__attribute__((constructor))` is not specific to the OS (e.g., Linux), it's specific to the **compiler** (e.g., gcc).  As such you can do it on both Windows or Linux but only with specific compilers that support `__attribute__`.  For example it wouldn't work with MSVC on the host side.

Reply:
The use of __attribute__((constructor)) has been removed.